bauhaus/core.py

```diff
@@ -150,15 +174,76 @@ def introspect(self):
                           "Try running compile() on your encoding.")
             return self.debug_constraints
 
+        print('\n\t{Encoding Introspection}\n')
+
         for constraint, clause in self.debug_constraints.items():
             print(f"{constraint}: \n")
-            if constraint.instance_constraints:
-                for instance, values in constraint.instance_constraints.items():
-                        print(f"{instance} =>")
-                        for v in values:
-                            print(f"{v}")
-                        print("\n")
-            print(f"Final {constraint._constraint.__name__}: {clause} \n")
+            # Check based on original constraint type
+            if 'instance_constraints' in dir(constraint):
+                if constraint.instance_constraints:
+                    for instance, values in constraint.instance_constraints.items():
+                            print(f"{instance} =>")
+                            for v in values:
+                                self.pprint(v)
+                            print("\n")
+                print(f"Final {constraint._constraint.__name__}: ", end='')
+                self.pprint(clause)
+            # Otherwise, it must be coming from a raw constraint
+            else:
+                print('Raw constraint added:')
+                self.pprint(clause)
+                print()
+
+        print()
+
+    def pprint(self, formula):
+        """Pretty print an NNF formula
+
+        Arguments
+        ---------
+        formula : NNF
+            Formula to be displayed.
+        """
+
+        def _process(f):
+            if isinstance(f, Var):
+                return {True: '', False: '¬'}[f.true] + str(f.name)
+            elif isinstance(f, And):
+                return '(' + ' ∧ '.join([_process(i) for i in f]) + ')'
+            elif isinstance(f, Or):
+                return '(' + ' ∨ '.join([_process(i) for i in f]) + ')'
+            else:
+                raise TypeError("Can only pprint an NNF object. Given %s" % type(f))
+
+        print(_process(formula))
+
+
+class CustomNNF:
```

Comment:
could we add a class description here to describe the purpose of CustomNNF/how it works and its attributes